### PR TITLE
Switch to Typescript and modern Javascript

### PR DESCRIPTION
**Pull request by lucsoft**

### CODE REVIEW

**Comment by 5saviahv** [comment]
I created `typescript` branch can you open PR against it so master is safe until it is ready.

**Comment by lucsoft** [comment]
Status: One test is failing and i don't know why

**Comment by lucsoft** [comment]
@5saviahv can you make a review to this pr?

File: package.json [hunk 1]
```diff
@@ -1,48 +1,56 @@
...
+        "@types/chai": "^4.2.16",
+        "@types/mocha": "^8.2.2",
+        "@types/rimraf": "^3.0.0",
+        "chai": "^4.3.4",
+        "mocha": "^8.3.2",
+        "prettier": "^2.2.1",
+        "rimraf": "^3.0.2",
+        "ts-node": "^9.1.1"
+    },
+    "dependencies": {
```
Comment (5saviahv) [inline package.json:52]: typescipt here will tell typescript is needed even after tsc compiling

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
+{
+    "compilerOptions": {
+        "incremental": true,
+        "target": "ESNext",
```
Comment (5saviahv) [inline tsconfig.json:4]: I dont think ESNext is compatible with Node 6 ???

File: util/index.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+export * from './utils';
+export * as FileSystem from './fileSystem';
+export { Constants } from './constants';
+export { Errors } from './errors';
+export * from './fattr';
```
Comment (5saviahv) [inline util/index.ts:5]: all fattr functions should stay in FileAttr object, I believe currently they are at top level

File: methods/zipcrypto.ts [hunk 4]
```diff
@@ -38,76 +36,77 @@ const genSalt = () => {
...
-// Class Initkeys handles same basic ops with keys
-function Initkeys(pw) {
-    const pass = Buffer.isBuffer(pw) ? pw : Buffer.from(pw);
-    this.keys = new Uint32Array([0x12345678, 0x23456789, 0x34567890]);
-    for (let i = 0; i < pass.length; i++) {
-        this.updateKeys(pass[i]);
+export class Initkeys {
+    keys: Uint32Array;
+
+    constructor(pw: any) {
```
Comment (5saviahv) [inline methods/zipcrypto.ts:50]: pw can be string (as utf8) or Buffer

File: adm-zip.ts [hunk 5]
```diff
@@ -0,0 +1,726 @@
...
+const defaultOptions = {
+    // option "noSort" : if true it disables files sorting
+    noSort: false,
+    // read entries during load (initial loading may be slower)
+    readEntries: false,
+    // default method is none
+    method: Constants.NONE
+};
+
+function canonical(p: string) {
```
Comment (5saviahv) [inline adm-zip.ts:17]: Maybe these functions should be moved to utils. This way we can write tests for these functions.

**Comment by 5saviahv** [comment]
I will squash and merge it. I simply like to make some commits here also

File: tsconfig.json (hunk 2 above)
Comment (lucsoft) [reply tsconfig.json:4]: The active LTS is 14 if you need node 6 or so babel is the way to go

File: package.json (hunk 1 above)
Comment (lucsoft) [reply package.json:52]: Yes it was just the default can be changed to dev dependencies

File: tsconfig.json (hunk 2 above)
Comment (5saviahv) [reply tsconfig.json:4]: I dont think we should racing with node versions, Instead we should look features what each version offers.

With typescript you can code with latest language features but support older versions as well. Typescript allows library use with Deno as well.

File: package.json (hunk 1 above)
Comment (5saviahv) [reply package.json:52]: I moved them into dev section in repo

File: tsconfig.json (hunk 2 above)
Comment (lucsoft) [reply tsconfig.json:4]: It's not really racing just using the active LTS as a target. And with new features we could trade compatibility with performance and readability

Deno doesn't support npm modules natively and the api we are using is probably also not available

Comment (5saviahv) [reply tsconfig.json:4]: Sorry I still dont get it, I would like to keep next npm versions compatible with previous adm-zip versions if it is not STRICTLY needed for some feature, but with typescript, anyone who wants can compile it for node 16 API.

Yes, Deno is not aim in any way.

**Comment by lucsoft** [comment]
@5saviahv  @cthackers Hey how its going with typescript for adm-zip the current main doesn't include any typescript

**Pull request merged**